### PR TITLE
New topo capabilities

### PR DESCRIPTION
**Pull request by gvallee**

### CODE REVIEW

File: include/dpu_offload_group_cache.h [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ * are numbered from 0 to the number of hosts involved in the group.
+ * 
+ * @param[in] engine Associated offload engine
+ * @param[in] group_uid Target group identified by its unique group identifier
+ * @param[in,out] host_id Host identifier from the ordered list of hosts that are involved in the group
+ * @return dpu_offload_status_t 
+ */
+dpu_offload_status_t get_ordered_host_id_by_group(offloading_engine_t *engine,
+                                                  group_uid_t group_uid,
+                                                  host_info_t *host_id);
```
Comment (bws) [inline include/dpu_offload_group_cache.h:72]: This should be a size_t (an array index into the host array)

Comment (gvallee) [reply include/dpu_offload_group_cache.h]: Fixed

**Review by bws — CHANGES_REQUESTED** [review]

File: include/dpu_offload_group_cache.h [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+ * the hosts that MIMOSA creates from the job configuration (in most
+ * cases, the order from the configuration file). All host in the list
+ * are numbered from 0 to the number of hosts involved in the group.
+ * 
+ * @param[in] engine Associated offload engine
+ * @param[in] group_uid Target group identified by its unique group identifier
+ * @param[in,out] host_idx Host index from the ordered list of hosts that are involved in the group.
+ * @return dpu_offload_status_t 
+ */
+dpu_offload_status_t get_ordered_host_id_by_group(offloading_engine_t *engine,
```
Comment (bws) [inline include/dpu_offload_group_cache.h:70]: This should be get_host_idx_by_group

File: include/dpu_offload_group_cache.h [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+ * @brief Get the number of service process on a specific host within a specific group.
+ * 
+ * @param[in] engine Associated offload engine
+ * @param[in] group_uid Target group identified by its unique group identifier
+ * @param[in] host_id Target host identifier
+ * @param[in,out] num_sps Returned number of service process on the target host;
+ * returns 0 is the target host is not involved in the group
+ * @return dpu_offload_status_t 
+ */
+dpu_offload_status_t get_num_sps_by_group_host(offloading_engine_t *engine,
```
Comment (bws) [inline include/dpu_offload_group_cache.h:84]: this can be get_num_sps_by_group_host_idx

File: include/dpu_offload_group_cache.h [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+ * @brief Get the number of ranks running on a target host within a specific group.
+ * 
+ * @param[in] engine Associated offload engine
+ * @param[in] group_uid Target group identified by its unique group identifier
+ * @param[in] host_id Target host within the group
+ * @param[in,out] num_ranks Returned number of ranks or 0 if the host is not involved in the group
+ * @return dpu_offload_status_t DO_SUCCESS or DO_ERROR if case of an error during execution of the function.
+ * If the host is not part of the group, it does not raise an exception.
+ */
+dpu_offload_status_t get_num_ranks_for_group_host(offloading_engine_t *engine,
```
Comment (bws) [inline include/dpu_offload_group_cache.h:115]: could be get_num_ranks_for_group_host_idx

File: include/dpu_offload_group_cache.h [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+ * 
+ * @param[in] engine Associated offload engine
+ * @param[in] group_uid Target group identified by its unique group identifier
+ * @param[in] host_id Target host identifier
+ * @param[in] rank Target rank in the group
+ * @param[in,out] idx Returned index associated to the rank from the array of ranks running on the host.
+ * If the rank is not running on the host or beyond the group size, an internal error is raised
+ * @return dpu_offload_status_t 
+ */
+dpu_offload_status_t get_n_for_rank_by_group_host(offloading_engine_t *engine,
```
Comment (bws) [inline include/dpu_offload_group_cache.h:131]: get_rank_idx_by_group_host_idx

File: include/dpu_offload_group_cache.h [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+ * @param[in] group_uid Target group identified by its unique group identifier
+ * @param[in] host_id Target host identifier
+ * @param[in,out] sps Pointer to the internal lookup array associated to the list of service processes for the target host.
+ * Since the table is exposed directly from the cache, the caller must not free the array, it will be freed when the group cache is freed.
+ * If no service process on the target host is involved in the group, the pointer is set to NULL and no exception is raised.
+ * @param[in,out] num_sps Number of service processes associated to the target host within the group;
+ * 0 if no service process on the host is involved in the group.
+ * @return dpu_offload_status_t 
+ */
+dpu_offload_status_t get_all_sps_by_group_host(offloading_engine_t *engine,
```
Comment (bws) [inline include/dpu_offload_group_cache.h:150]: get_all_sps_by_group_host_idx

File: include/dpu_offload_group_cache.h [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+/**
+ * @brief Get the n^th service process on a given host that is involved in the target group.
+ * 
+ * @param[in] engine Associated offload engine
+ * @param[in] group_uid Target group identified by its unique group identifier
+ * @param[in] host_id Target host identifier
+ * @param[in,out] sp_group_gid Global group identifier of the target service process
+ * @return dpu_offload_status_t 
+ */
+dpu_offload_status_t get_nth_sp_by_group_host(offloading_engine_t *engine,
```
Comment (bws) [inline include/dpu_offload_group_cache.h:204]: get_local_sp_idx_by_group_host_idx

File: include/dpu_offload_group_cache.h [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+ * @param[in] sp_group_gid Global group service process identifer, as for example returned by get_local_sp_id_by_group(). Must be valid.
+ * @param[in,out] ranks Pointer to the internal lookup array associated to the list of ranks associated with a given service process.
+ * The caller must not free the array at any point since it is an internal array directly exposed to users. It is freed when
+ * the group cache is being freed.
+ * @param[in,out] num_ranks Number of ranks associated to the target service process. The returned value must strictly be greater than
+ * 0 since the global group service identifier is assumed valid and therefore a strict minimum of one rank must be associated to the
+ * service process (otherwise the global group service process would be invalid)
+ * @return dpu_offload_status_t 
+ */
+dpu_offload_status_t get_all_ranks_by_group_sp(offloading_engine_t *engine,
```
Comment (bws) [inline include/dpu_offload_group_cache.h:213]: I realize I need this to be a host_idx with a local sp_id rather than a global sp_id

**Pull request merged**